File: Ahorn/entities/CustomTorch.jl
module VivHelperCustomTorch

using ..Ahorn, Maple
using Ahorn.VivHelper

@mapdef Entity "VivHelper/CustomTorch" CustomTorch(
	x::Integer,
	y::Integer,
	startLit::Bool=false,
	Color::String="80ffff",
	spriteColor::String="80ffff",
	Alpha::Number=1.0,
	startFade::Integer=48,
	endFade::Integer=64,
	RegisterRadius::Number=4.0,
	unlightOnDeath::Bool=false
)

const placements = Ahorn.PlacementDict(
	"Colored Torch (Viv's Helper)" => Ahorn.EntityPlacement(
		CustomTorch
	)
)
Ahorn.editingOptions(entity::CustomTorch) = Dict{String, Any}(
    "Color" => VivHelper.XNAColors
)

function torchSprite(entity::CustomTorch)
	lit = get(entity.data, "startLit", false)

	return lit ? "ahorn/VivHelper/torch/grayTorchLit.png" : "ahorn/VivHelper/torch/grayTorchUnlit.png"
end

function Ahorn.selection(entity::CustomTorch)
	x, y = Ahorn.position(entity)
	sprite = torchSprite(entity)

	return Ahorn.getSpriteRectangle(sprite, x, y)
end

function Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::CustomTorch, room::Maple.Room)
	sprite = torchSprite(entity)
	tint1 = VivHelper.ColorFix(get(entity.data, "spriteColor", "ffffff"), 1.0)
	Ahorn.drawSprite(ctx, sprite, 0, 0, tint=tint1)
end

end

File: Loenn/entities/CustomTorch.lua
local vivUtil = require('mods').requireFromPlugin('libraries.vivUtil')

local torch = {}

torch.name = "VivHelper/CustomTorch"
torch.depth = 2000
torch.placements = {
    name = "main",
    data = {
        startLit=false,
	    color="80ffff",
	    spriteColor="80ffff",
	    Alpha=1.0,
	    startFade=48,
	    endFade=64,
	    RegisterRadius=4.0,
	    unlightOnDeath=false
    }
}
torch.fieldInformation = {
    color = {fieldType = "color", allowXNAColors = true, useAlpha=true },
    spriteColor = {fieldType = "color", allowXNAColors = true, useAlpha=true },
    Alpha = {fieldType = "number", minimumValue = 0.0, maximumValue = 1.0},
    startFade = {fieldType = "integer", minimumValue = 0, maximumValue = 120 },
    endFade = {fieldType = "integer", minimumValue = 0, maximumValue = 120 }, -- Fun fact, this is the actual limit for lights in celeste
    RegisterRadius = {fieldType = "number", minimumValue = 2}
}

function torch.texture(room, entity)
    return entity.startLit and "ahorn/VivHelper/torch/grayTorchLit" or "ahorn/VivHelper/torch/grayTorchUnlit"
end

function torch.color(room, entity)
    return vivUtil.GetColorTable(entity.spriteColor, true, {1,1,1,1})
end

return torch